## About the Browser

This browser allows the user to search for a gene by the HUGO gene symbol and 
dynamically returns information about its expression. Gene expression data were 
generated by [Lappalainen et al.](https://www.nature.com/articles/nature12531.pdf) 
for a subset of the 1000 Genomes Project population samples. 

The Table tab of the Browser page displays a table containing gene expression 
information for the selected samples with links to repository samples. This table
can be downloaded by the user by clicking the 'Download' button beneath the table. 
The Expression Histogram tab of the Browser page displays an interactive histogram 
of the gene expression data across the selected samples. Individual samples can be 
highlighted on the histogram by typing or selecting their sample ID in the dropdown. 
The About page contains information about data collection and processing.

## Definitions:

- Fragments per kilobase per million (FPKM)
- Transcripts per kilobase per million (TPM)
- Counts per million (CPM)
- Reads per kilobase per million (RPKM)
- Variance stabilizing transformation (VST)

## Data Processing:

Preprocessed [TPM](https://www.ebi.ac.uk/gxa/experiments-content/E-GEUV-1/resources/ExperimentDownloadSupplier.RnaSeqBaseline/tpms.tsv) 
and [FPKM](https://www.ebi.ac.uk/gxa/experiments-content/E-GEUV-1/resources/ExperimentDownloadSupplier.RnaSeqBaseline/fpkms.tsv) 
data were downloaded directly from EBI Expression Atlas and integrated into the 
application. Please see the experiment [supplementary information](https://www.ebi.ac.uk/gxa/experiments/E-GEUV-1/Supplementary%20Information)
for details about FPKM and TPM processing.

For CPM, RPKM, and VST normalized data, raw gene counts and annotation information were 
downloaded from [Array Express](https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/)
for the E-GEUV-1 dataset. For more information about the generation of the raw count
data see the [E-GEUV Array Express](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEUV-1/)
page for details and links to the associated publications. Raw counts were summed
across biological replicates and then normalized using TMM normalization in 
[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html). CPM and RPKM
normalized values were generated using the `cpm()` and `rpkm()` functions in `edgeR`, respectively.
Gene length information required in the RPKM calculation was downloaded from [GENCODE v12](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_12/gencode.v12.annotation.gtf.gz).
VST normalized data were calculated on the summed count matrix using the [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) package.
